book/src/solana-scripts.md

```diff
@@ -0,0 +1,108 @@
+# Solana Scripts
+
+This architecture describes how the Pipeline runtime supports scripts. 
+
+In Solana programs have absolute control over the state assigned to them.  The
+goal of this architecture is provide a simple way for on-chain programs to
+interact with other programs without violating the state ownership guarantees.
+
+Contract composition has led to multiple multi-million dollar bugs in Ethereum
+
+* https://www.parity.io/security-alert-2/
+
+## Scripts
+
+The purpose of scripts is to execute logic between instructions.
+
+For example, assignment of tokens to the winner based on an output of a game can
+be defined as a script.
+
+### `Loader::CreateScript`
+
+`Loader::CreateScript` designates that the loaded executable bytecode is a
+script. The difference between scripts and programs is that script execution
+yields to external process instructions.
```

Comment:
In the issue I created, I call this a "trap", suggesting that one traps into the kernel for a synchronous call to `process_instruction()`. You say "yield", suggesting saving a continuation for some asynchronous call and then go on with an example that calls `send_message()` further suggesting an asynchronous call. I don't see the value in that over trapping to the runtime for a synchronous call.

Reply:
So the trap still needs to guarantee that the same rules we enforce for instructions are enforced for the trapped call.

I think that dynamic calls are dangerous. The user should explicitly authorize a specific path of execution for the script. Then any transfers of value are explicit and deterministic.